ng serve no longer works without AoT

### Command

serve

### Is this a regression?

- [X] Yes, this behavior used to work in the previous version

### The previous version in which this bug was not present was

16

### Description

When using `"aot": false`, there are two critical issues with `ng serve`.

1. `.html` files do not seem to be watched, so changes do not trigger a rebuild/refresh.
2. Saving `.ts` files trigger a rebuild, but the build always fail (see error below).

### Minimal Reproduction

1. `ng new myproject`
2. `cd myproject`
3. Add `"aot": true` to the angular.json builder options (`projects.myproject.architect.build.options.aot`)
4. `ng serve`
5. Make changes to and save `app.component.html` and `app.component.ts`

### Exception or Error

```text
Two issues:

1. Changes to `app.component.html` does not trigger rebuild. The only way for the changes to take effect is to quit and run `ng serve` again.


2. When saving `app.component.ts`, an error is shown:
⠋ Changes detected. Rebuilding...
✘ [ERROR] File 'src/app/app.component.ts' is missing from the TypeScript compilation. [plugin angular-compiler]

    src/main.ts:3:29:
      3 │ import { AppComponent } from './app/app.component';
        ╵                              ~~~~~~~~~~~~~~~~~~~~~

  Ensure the file is part of the TypeScript program via the 'files' or 'include' property.
```


### Your Environment

```text
_                      _                 ____ _     ___
    / \   _ __   __ _ _   _| | __ _ _ __     / ___| |   |_ _|
   / △ \ | '_ \ / _` | | | | |/ _` | '__|   | |   | |    | |
  / ___ \| | | | (_| | |_| | | (_| | |      | |___| |___ | |
 /_/   \_\_| |_|\__, |\__,_|_|\__,_|_|       \____|_____|___|
                |___/
    

Angular CLI: 17.0.6
Node: 20.9.0
Package Manager: npm 10.2.3
OS: darwin arm64

Angular: 17.0.6
... animations, cli, common, compiler, compiler-cli, core, forms
... platform-browser, platform-browser-dynamic, router

Package                         Version
---------------------------------------------------------
@angular-devkit/architect       0.1700.6
@angular-devkit/build-angular   17.0.6
@angular-devkit/core            17.0.6
@angular-devkit/schematics      17.0.6
@schematics/angular             17.0.6
rxjs                            7.8.1
typescript                      5.2.2
zone.js                         0.14.2
```


### Anything else relevant?

This happens both with the `browser-eslint` and the `application` builder.

--- a/packages/angular_devkit/build_angular/src/builders/application/tests/behavior/typescript-rebuild-touch-file_spec.ts
+++ b/packages/angular_devkit/build_angular/src/builders/application/tests/behavior/typescript-rebuild-touch-file_spec.ts
@@ -1,0 +1,52 @@
+/**
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import { concatMap, count, take, timeout } from 'rxjs';
+import { buildApplication } from '../../index';
+import { APPLICATION_BUILDER_INFO, BASE_OPTIONS, describeBuilder } from '../setup';
+
+describeBuilder(buildApplication, APPLICATION_BUILDER_INFO, (harness) => {
+  describe('Behavior: "Rebuilds when touching file"', () => {
+    for (const aot of [true, false]) {
+      it(`Rebuild correctly when file is touched with ${aot ? 'AOT' : 'JIT'}`, async () => {
+        harness.useTarget('build', {
+          ...BASE_OPTIONS,
+          watch: true,
+          aot,
+        });
+
+        const buildCount = await harness
+          .execute({ outputLogsOnFailure: false })
+          .pipe(
+            timeout(30_000),
+            concatMap(async ({ result }, index) => {
+              switch (index) {
+                case 0:
+                  expect(result?.success).toBeTrue();
+                  // Touch a file without doing any changes.
+                  await harness.modifyFile('src/app/app.component.ts', (content) => content);
+                  break;
+                case 1:
+                  expect(result?.success).toBeTrue();
+                  await harness.removeFile('src/app/app.component.ts');
+                  break;
+                case 2:
+                  expect(result?.success).toBeFalse();
+                  break;
+              }
+            }),
+            take(3),
+            count(),
+          )
+          .toPromise();
+
+        expect(buildCount).toBe(3);
+      });
+    }
+  });
+});

--- a/packages/angular_devkit/build_angular/src/tools/esbuild/angular/source-file-cache.ts
+++ b/packages/angular_devkit/build_angular/src/tools/esbuild/angular/source-file-cache.ts
@@ -31,7 +31,6 @@ export class SourceFileCache extends Map<string, ts.SourceFile> {
     }
     for (let file of files) {
       file = path.normalize(file);
-      this.typeScriptFileCache.delete(file);
       this.loadResultCache.invalidate(file);
 
       // Normalize separators to allow matching TypeScript Host paths

